### PR TITLE
Feat/2326 content arranger

### LINKED ISSUE #2326: Split chat entries into dialogs by discussing topics

### PR DESCRIPTION
**Pull request by crui3er**
Closes #2326

### CODE REVIEW

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        }
+
+        foreach (var builder in builders) {
+            if (builder.HasModified)
+                yield return new SourceEntries(null, null, builder.Entries);
+        }
+    }
+
+    private async ValueTask<IReadOnlyList<ChatEntry>> LoadTailEntries(
```
Comment (AlexUstinov) [inline src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs:133]: For now, this method and also the rest of the code in this class do not respect the fact one entry could be split into multiple documents. We should plan to implement this functionality.

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArrangerSelector.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using ActualChat.Users;
+
+namespace ActualChat.MLSearch.Indexing.ChatContent;
+
+internal interface IChatContentArrangerSelector
+{
+    Task<IChatContentArranger> GetContentArranger(ChatId chatId);
+}
+
+internal sealed class ChatContentArrangerSelector(IServiceProvider services) : IChatContentArrangerSelector
```
Comment (AlexUstinov) [inline src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArrangerSelector.cs:11]: Can we use direct dependency on `IServerKvasBackend` ?

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArrangerSelector.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public async Task<IChatContentArranger> GetContentArranger(ChatId chatId)
+    {
+        var useArranger2 = false;
+        var chatSidList = await ServerSettingsKvas.Get<string>(Constants.ServerSettings.UseChatContentArranger2ChatIds).ConfigureAwait(false);
+        if (!chatSidList.IsNullOrEmpty()) {
+            var chatSids = chatSidList.Split(';');
+            useArranger2 = chatSids.Any(c => OrdinalEquals(chatId.Id.Value, c));
+        }
+
+        IChatContentArranger contentArranger = useArranger2
```
Comment (AlexUstinov) [inline src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArrangerSelector.cs:26]: We can resolve arrangers as a collection. So, with a small addition to the `IChatContentArranger` interface, that adds a property returning arranger Id, we can easily build arranger dictionary. Using dictionary for resolves eliminates the need of `IServiceProvider` dependency.

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs [hunk 4]
```diff
@@ -0,0 +1,150 @@
+using ActualChat.Chat;
+using ActualChat.MLSearch.Documents;
+
+namespace ActualChat.MLSearch.Indexing.ChatContent;
+
+internal sealed class ChatContentArranger2(
+    IChatsBackend chatsBackend,
+    IDialogFragmentAnalyzer fragmentAnalyzer,
+    ChatDialogFormatter chatDialogFormatter
```
Comment (AlexUstinov) [inline src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs:9]: It makes sense to introduce `IChatDialogFormatter` interface for this dependency.

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/DialogFragmentAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private readonly Dictionary<string, PromptTemplate> _promptTemplates = new(StringComparer.Ordinal);
+
+    private ILogger Log { get; } = log;
+    private bool IsDiagnosticsEnabled => options.IsDiagnosticsEnabled;
+
+    public async Task<int> ChooseOption(string phrase, string[] fragments)
+    {
+        var sb = new StringBuilder();
```
Comment (AlexUstinov) [inline src/dotnet/MLSearch.Service/Indexing/ChatContent/DialogFragmentAnalyzer.cs]: As we actively use `StrilngBuilder` instances here it makes sense to have a pool of string builders, so we could reduce number of allocations.

**Review by AlexUstinov — APPROVED** [review]

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs (hunk 4 above)
Comment (crui3er) [reply src/dotnet/MLSearch.Service/Indexing/ChatContent/ChatContentArranger2.cs:9]: Why?

File: src/dotnet/MLSearch.Service/Indexing/ChatContent/DialogFragmentAnalyzer.cs (hunk 5 above)
Comment (crui3er) [reply src/dotnet/MLSearch.Service/Indexing/ChatContent/DialogFragmentAnalyzer.cs]: Done

**Pull request merged**